runtime/src/utils/buffer/write.rs

```diff
@@ -176,153 +210,97 @@ impl<B: Blob> Blob for Write<B> {
         buf: impl Into<StableBuf> + Send,
         offset: u64,
     ) -> Result<StableBuf, Error> {
-        // Acquire a read lock on the inner state
-        let inner = self.inner.read().await;
-
-        // Ensure offset read doesn't overflow
+        // Prepare `buf` to capture the read data.
         let mut buf = buf.into();
-        let data_len = buf.len();
-        let data_end = offset
-            .checked_add(data_len as u64)
+        let buf_len = buf.len(); // number of bytes to read
+
+        // Ensure the read doesn't overflow.
+        let end_offset = offset
+            .checked_add(buf_len as u64)
             .ok_or(Error::OffsetOverflow)?;
 
-        // If the data required is beyond the buffer end, return an error
-        let buffer_start = inner.position;
-        let buffer_end = buffer_start + inner.buffer.len() as u64;
+        // Acquire a read lock on the buffer.
+        let buffer = self.buffer.read().await;
 
-        // Ensure we don't read beyond the logical end of the blob
-        if data_end > buffer_end {
+        // If the data required is beyond the size of the blob, return an error.
+        if end_offset > buffer.size() {
             return Err(Error::BlobInsufficientLength);
         }
 
-        // Case 1: Read entirely from the underlying blob (before buffer)
-        if data_end <= buffer_start {
-            return inner.blob.read_at(buf, offset).await;
-        }
-
-        // Case 2: Read entirely from the buffer
-        if offset >= buffer_start {
-            let buffer_offset = (offset - buffer_start) as usize;
-            let end_offset = buffer_offset + data_len;
-
-            if end_offset > inner.buffer.len() {
-                return Err(Error::BlobInsufficientLength);
-            }
-
-            buf.put_slice(&inner.buffer[buffer_offset..end_offset]);
+        // Read any bytes from the buffer that overlap with the requested range.
```

Comment:
May be worth emphasizing any remaining is at the beginning here as well.

Reply:
Added this emphasis right before "blob_part" is read.